tests/unit/test_shared.py

```diff
@@ -0,0 +1,126 @@
+    with pytest.raises(ValidationError, match='Rule name .* does not match ".*"'):
+        RuleDbModelFactory.build(rules={"test": RuleModelFactory.build(name="invalid")})
+
+
+def test_validate_rule_name_with_no_rules() -> None:
+    """Test rules db with no rules."""
+    RuleDbModelFactory.build()
+
+
+def test_validate_multiple_defaults() -> None:
+    """Should raises if multiple default rule."""
+    with pytest.raises(ValidationError, match="Only one default rule is allowed"):
+        RuleDbModelFactory.build(
+            default={
+                "rule-1": RuleModelFactory.build(name="rule-1"),
+                "rule-2": RuleModelFactory.build(name="rule-2"),
+            }
+        )
+
+
+def test_validate_default_invalid_name() -> None:
+    """Should raises if default rule as an invalid name."""
+    with pytest.raises(
+        ValidationError, match='Default rule name ".*" does not match ".*"'
+    ):
+        RuleDbModelFactory.build(
+            default={
+                "rule-1": RuleModelFactory.build(),
+            }
+        )
+
+
+def test_validate_all_in_category_with_valid_category() -> None:
+    """Test rule with valid category."""
+    RuleModelFactory.build(categories=["ALL"])
+
+
+def test_validate_all_in_category_with_invalid_category() -> None:
+    """Should raises if ALL is missing in categories."""
+    with pytest.raises(ValidationError, match='Rule .* is missing category "ALL"'):
+        RuleModelFactory.build(categories=["invalid"])
+
+
+def test_validate_description_length_with_valid_description() -> None:
+    """Test rule with valid description."""
+    RuleModelFactory.build(description="Lorem Ipsum " * 42)
+
+
+def test_validate_description_length_with_invalid_description() -> None:
+    """Should raises if rule description is too long."""
+    with pytest.raises(
+        ValidationError,
+        match="Rule .* has a description longer than 512 characters",
+    ):
+        RuleModelFactory.build(description="Lorem Ipsum " * 43)
+
+
+@pytest.mark.parametrize("url", ["https://example.com", "http://example.com"])
+def test_validate_ref_url_with_valid_url(url: str) -> None:
+    """Test rule with valid url."""
+    rule = RuleModelFactory.build(ref=url)
+    assert url == rule.ref
+
+
+def test_validate_ref_url_with_invalid_url() -> None:
+    """Should raises if ref is not a valid url."""
+    with pytest.raises(
+        ValidationError,
+        match='Url missing protocol: "invalid_url" from rule ".*"',
+    ):
+        RuleModelFactory.build(ref="invalid_url")
+
+
+def test_validate_ref_url_with_valid_url_with_placeholder(
+    monkeypatch: MonkeyPatch,
+) -> None:
+    """Test rule ref with env var injection."""
+    env_var_name = "BOOSTSEC_DOC_BASE_URL"
+    monkeypatch.setenv(env_var_name, "http://test.com")
+    rule = RuleModelFactory.build(ref=f"{{{env_var_name}}}/a/b/c")
+    assert rule.ref == "http://test.com/a/b/c"
+
+
+def test_render_doc_url(monkeypatch: MonkeyPatch) -> None:
```

Comment:
(optional) Since refactoring, maybe matching the location of the tested method in the tested would be good

Reply:
Yeah, now that I think of it, I don't think we need this anymore. The "render" will be done when init the model, so there's no need to call that function ever again.